core/src/listener/state.ts

```diff
@@ -179,31 +175,25 @@ export class State {
     const contractAddress = toAddListener.address
     const observedBlockRedisKey = getObservedBlockRedisKey(contractAddress)
     const latestBlock = await this.latestBlockNumber()
+    const { blockKey: observedBlockKey } = await getObservedBlock({
+      blockKey: observedBlockRedisKey,
+      logger: this.logger
+    })
 
-    switch (this.listenerInitType) {
-      case 'clear':
-        // Clear metadata about previously observed blocks for a specific
-        // `contractAddress`.
-        await this.redisClient.set(observedBlockRedisKey, latestBlock - 1)
-        break
-
-      case 'latest':
-        await this.setObservedBlockNumberIfNotDefined(observedBlockRedisKey, latestBlock - 1)
-        break
-
-      default:
-        // [block number] initialization
-        await this.setObservedBlockNumberIfNotDefined(observedBlockRedisKey, latestBlock - 1)
-        for (let blockNumber = this.listenerInitType; blockNumber < latestBlock; ++blockNumber) {
-          const historyOutData: IHistoryListenerJob = {
-            contractAddress,
-            blockNumber
-          }
-          await this.historyListenerQueue.add('history', historyOutData, {
-            ...LISTENER_JOB_SETTINGS
-          })
-        }
-        break
+    /**
+     when listener starts, there are two options:
+      * latest observedBlock (key) exists -> do nothing 
+        (start latest job, it'll handle multiple blocks 
+        between observedBlock and latestBlock)
+      * it does not exist -> upsert latestBlock
+     */
+    if (observedBlockKey === '') {
```

Comment:
How would we handle situation where we deploy this new listener, but while we are deploying we do not listen on new blocks and some block are going to be missed?

Reply:
this scenario, `if (observedBlockKey === '')` happens only one time when we migrate to this new logic, where the `observedBlockKey` key wont exist in the db and it will be upserted with latest block number. Consequent service restarts will always have observedBlock number and all unprocessed blocks will be passed to the latestQ

There is a reason why I decided to pass unprocessed blocks to latestQ instead of historyQ but now that I think about it again, it should be totally fine. I'll reimplement passing unprocessed blocks to historyQ.

Regarding one time migration, it'd be ideal to figure out the observedBlock number from currently running listener service redis instance and initialize the observedBlock key value pair in the db before the new listener service is started. I'll look into it before we deploy.